src/v/iceberg/rest_client/types.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/sstring.hh>
+
+#include <ada/expected.h>
+
+namespace iceberg::rest_client {
+
+struct json_parse_error {
+    ss::sstring context;
+    ss::sstring error;
+};
+
+// The domain error represents the sum of all error types which can be
+// encountered during rest-client operations.
+using domain_error = std::variant<ss::sstring, json_parse_error>;
+
+// The core result type used by all operations in the iceberg/rest-client which
+// can fail. Allows chaining of operations together and short-circuiting when an
+// earlier operation in the chain has failed.
+template<typename T>
+using expected = tl::expected<T, domain_error>;
+
+// Oauth token returned by the catalog server, in exchange for credentials
+struct oauth_token {
+    ss::sstring token;
+    ss::lowres_clock::duration expiry;
+    ss::lowres_clock::time_point expires_at;
```

Comment:
The expiry is part of the token returned by the endpoint, we need `expires_at` for comparing the current time point with token expiry. We probably can do without storing the expiry. I'll see if it can be removed.

Reply:
removed this field - it made testing easier but the test can be adjusted